source/common/http/conn_manager_impl.cc

```diff
@@ -201,10 +201,10 @@ StreamDecoder& ConnectionManagerImpl::newStream(StreamEncoder& response_encoder)
   new_stream->response_encoder_ = &response_encoder;
   new_stream->response_encoder_->getStream().addCallbacks(*new_stream);
   new_stream->buffer_limit_ = new_stream->response_encoder_->getStream().bufferLimit();
-  // Make sure new streams are apprised that the underlying connection is blocked.
-  if (read_callbacks_->connection().aboveHighWatermark()) {
-    new_stream->callHighWatermarkCallbacks();
-  }
+  // If the network connection is backed up, the stream should be made aware of
+  // it on creation. Both HTTP/1.x and HTTP/2 codecs handle this.
+  ASSERT(read_callbacks_->connection().aboveHighWatermark() == false ||
+         new_stream->high_watermark_count_ > 0);
```

Comment:
I'm confused when/how `high_watermark_count_` gets incremented in this call chain in order to make this assert true. From looking at the HTTP/2 codec it looks like we run high watermark callbacks in `ServerConnectionImpl::onBeginHeaders` before we call `newStream`. Wouldn't we have to run the callbacks instead when we call `addCallbacks` for this to work properly? I'm probably missing something here. If so can you add more comments?

Reply:
If I can interject, the callback are in fact run in `addCallbacks`.

Reply:
Ah yes, I see it, thanks. Can we potentially clarify the comment? This makes sense. I will take a look at the tests now.